src/azure-cli-core/azure/cli/core/command_recommender.py

```diff
@@ -127,86 +130,129 @@ def _set_aladdin_recommendations(self):
+                    'description': result['description'],
+                    'link': result['link']
+                }
+                recommendations.append(recommendation)
 
         self.aladdin_recommendations.extend(recommendations)
 
-    def recommend_a_command(self):
-        """Recommend a command for user when user's command fails.
-        The recommended command will be the best matched one from
-        both the help files and the aladdin recommendations.
+    def provide_recommendations(self):
+        """Provide recommendations either from Aladdin service or CLI help examples,
+        which include both commands and reference links along with their descriptions.
         """
-
-        def get_sorted_candidates(target_args, candidate_args_list):
-            """Get the sorted candidates by target arguments"""
+        from azure.cli.core.style import Style
+
+        def sort_recommendations(recommendations):
+            """Sort the recommendations by parameter matching.
+            The sorting rules below are applied in order:
+                1. Commands starting with the user's input command name are ahead of those don't
+                2. Commands having more matched arguments are ahead of those having less
+                3. Commands having less arguments are ahead of those having more
+            """
 
             candidates = []
-            for index, candidate_args in enumerate(candidate_args_list):
+            target_arg_list = self._normalize_parameters(self.parameters)
+            for recommendation in recommendations:
                 matches = 0
-                for arg in candidate_args:
-                    if arg in target_args:
-                        matches += 1
+                arg_list = self._normalize_parameters(recommendation['command'].split(' '))
+
+                # ignore commands that do not start with the use's input command name
+                if recommendation['command'].startswith('az {}'.format(self.command)):
+                    for arg in arg_list:
+                        if arg in target_arg_list:
+                            matches += 1
+                else:
+                    matches = -1
+
                 candidates.append({
-                    'candidate_args': candidate_args,
-                    'index': index,
+                    'recommendation': recommendation,
+                    'arg_list': arg_list,
                     'matches': matches
                 })
 
             # sort the candidates by the number of matched arguments and total arguments
-            candidates.sort(key=lambda item: (item['matches'], -len(item['candidate_args'])), reverse=True)
+            candidates.sort(key=lambda item: (item['matches'], -len(item['arg_list'])), reverse=True)
+
+            return [candidate['recommendation'] for candidate in candidates]
 
-            return candidates
+        def decorate_command(raw_command):
```

Comment:
The functionality for colorization can be extracted and put in `style.py` as a common util.

Reply:
done.